### PR TITLE
ApiV4: PUT /users/{user_id}/mfa

### PR DESCRIPTION
**Pull request by andreistanciu24**
Please make sure you've read the [pull request](http://docs.mattermost.com/developer/contribution-guide.html#preparing-a-pull-request) section of our [code contribution guidelines](http://docs.mattermost.com/developer/contribution-guide.html).

When filling in a section please remove the help text and the above text.

#### Summary
Update a user's MFA.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-5243

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs)
- [x] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

**Comment by grundleborg** [comment]
@andreistanciu24 Jenkins is reporting GOFMT issues in `model/client4.go`.

In case you didn't know, you can run `make check-style` locally to identify these formatting/linting errors before you commit your code.

**Comment by andreistanciu24** [comment]
Hey, thanks. I know about it but it seem I forgot to run it this time. Because I have some strange errors that don't allow me to run tests locally. Maybe you can help me out. Every time I run a test with go test -v -run='SomeTest' ./api4, even if I remove all my code, i get: 

# testmain
/tmp/go-build093117423/github.com/mattermost/platform/api4/_test/_testmain.go:154: cannot use matchString (type func(string, string) (bool, error)) as type testing.testDeps in argument to testing.MainStart:
	func(string, string) (bool, error) does not implement testing.testDeps (missing MatchString method)
FAIL	github.com/mattermost/platform/api4 [build failed]

Do you have any ideas?

**Comment by grundleborg** [comment]
@andreistanciu24 couple of questions to try and figure out what's wrong:

1. What version of go are you using?
2. Do you have any local changes, or are you running on a clean checkout of this PR's branch?
3. Do you get the same error when running the same test command on master?

**Comment by andreistanciu24** [comment]
I'm using go version go1.7.4. I get the same error when running the command on master.

**Comment by grundleborg** [comment]
Very strange. Shall we carry on this conversation on [Mattermost in the Developers channel](https://pre-release.mattermost.com/core/channels/developers)?

File: api4/user.go [hunk 1]
```diff
@@ -550,3 +552,65 @@ func getAudits(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		}
+
+		if err := app.SendMfaChangeEmail(user.Email, activate, user.Locale, c.GetSiteURL()); err != nil {
+			l4g.Error(err.Error())
+		}
+	}()
+
+	rdata := map[string]string{}
+	rdata["status"] = "ok"
+	w.Write([]byte(model.MapToJson(rdata)))
```
Comment (jwilander) [inline api4/user.go]: Can you use `ReturnStatusOK` here?

File: api4/user.go [hunk 2]
```diff
@@ -550,3 +552,65 @@ func getAudits(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		var err *model.AppError
+		if user, err = app.GetUser(c.Session.UserId); err != nil {
+			l4g.Warn(err.Error())
+			return
+		}
+
+		if err := app.SendMfaChangeEmail(user.Email, activate, user.Locale, c.GetSiteURL()); err != nil {
+			l4g.Error(err.Error())
+		}
+	}()
```
Comment (jwilander) [inline api4/user.go]: Would you be able to create an app layer function in `app/user.go` that will take in the `useId`, `token` and `activate`. It should perform the logic on whether to activate or deactivate MFA and it should contain the go routine that sends the email.

The audits can stay in the api layer

File: api4/user.go (hunk 1 above)
Comment (andreistanciu24) [inline api4/user.go]: done

File: api4/user.go (hunk 2 above)
Comment (andreistanciu24) [inline api4/user.go]: Makes more sense indeed.
Done.

File: api4/user_test.go [hunk 3]
```diff
@@ -924,3 +925,38 @@ func TestGetAudits(t *testing.T) {
...
+	_, resp := Client.UpdateMfa(ruser.Id, "12334", true)
+	CheckUnauthorizedStatus(t, resp)
+
+	Client.Login(user.Email, user.Password)
+	_, resp = Client.UpdateMfa("fail", "56789", false)
+	CheckBadRequestStatus(t, resp)
+
+	_, resp = Client.UpdateMfa(ruser.Id, "", true)
+	CheckErrorMessage(t, resp, "api.context.invalid_body_param.app_error")
+}
```
Comment (jwilander) [inline api4/user_test.go]: I don't see a success case being tested, can you add one?

Comment (andreistanciu24) [inline api4/user_test.go]: I tried to add one, but it seems that, even though I enabled MFA in the config keys, there is this message: "MFA not configured or available on this server" which also shows up in v3 test logs (which have no success case btw). 
I tried to dig into it and can't figure out why the MfaInterface is nil.  
Any hint on how to pass this issue?

Comment (jwilander) [inline api4/user_test.go]: Same way as I showed you for the editting post stuff, you need to temporarily set the license and config settings

Comment (andreistanciu24) [inline api4/user_test.go]: I tried that. Here's what I've got:
	enableMfa := *utils.Cfg.ServiceSettings.EnableMultifactorAuthentication
	defer func() {
		utils.IsLicensed = false
		*utils.License.Features.MFA = false
		*utils.Cfg.ServiceSettings.EnableMultifactorAuthentication = enableMfa
	}()
        ....
	utils.IsLicensed = true
	*utils.License.Features.MFA = true
	*utils.Cfg.ServiceSettings.EnableMultifactorAuthentication = true
--success test here--

**Comment by grundleborg** [comment]
@andreistanciu24 would you be able to rebase this branch onto current `master` so that the tests will run successfully on build.mattermost.com?

**Comment by jwilander** [comment]
@andreistanciu24 can you give this a rebase?

**Comment by andreistanciu24** [comment]
@jwilander  I did rebased with master but now it seems that it crasesh & stops right after TestGetIncomingWebhooks.
Until now it was my new test that was failing. 
Problems on master or Jenkins?

**Comment by jwilander** [comment]
Does it pass locally? Also can you squash your commits into one?

**Comment by andreistanciu24** [comment]
@jwilander  I just rebased, and run make test on local:
--- PASS: TestUserUploadProfileImage (2.56s)
=== RUN   TestUserUpdate
[2017/02/28 14:33:30 EET] [DEBG] /api/v3/teams/create
panic: test timed out after 10m50s

**Comment by jwilander** [comment]
If you rebase that should fix the timeout error since a recently merged PR just increased the timeout.

**Comment by andreistanciu24** [comment]
@jwilander  If you refer to this PR "Move webhook logic from api layer to app layer (#5527)", well, it didn't fix the timeout case. It's on this branch but the build still fails.

**Comment by jwilander** [comment]
Well it needs a rebase either way

**Comment by jwilander** [comment]
You've got a build failure:
```
runtime/debug.Stack(0xc422013701, 0xc4225e7940, 0xc4226c7cf0)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x79
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x22
github.com/mattermost/platform/api4.CheckNoError(0xc422398780, 0xc4225e7940)
	github.com/mattermost/platform/api4/_test/_obj_test/apitestlib.go:428 +0x71
github.com/mattermost/platform/api4.TestUpdateMfa(0xc422398780)
	/home/ubuntu/workspace/mattermost-platform-pr@2/src/github.com/mattermost/platform/api4/user_test.go:1071 +0x65b
testing.tRunner(0xc422398780, 0xc56b00)
	/usr/local/go/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:646 +0x2ec
--- FAIL: TestUpdateMfa (1.76s)
	apitestlib.go:429: Expected no error, got : MFA not configured or available on this server, 
```

**Comment by andreistanciu24** [comment]
Hi @jwilander  This is the case I've been mentioning you since a few days now: "I can't by-pass that MFA configuration for testing".

**Comment by jwilander** [comment]
Ok, I can pull the branch and take a look

**Pull request closed without merging**